docs/rfds/per-request-cancellation.mdx

```diff
@@ -0,0 +1,258 @@
+
+## Elevator pitch
+
+Introduce a standardized per-request cancellation mechanism for the Agent Client Protocol, inspired by the Language Server Protocol (LSP), to enable a more granular cancellation of requests where individual JSON-RPC requests can be cancelled one by one.
+
+## Status quo
+
+The JSON-RPC specification doesn't define any standard mechanism for request cancellation and leaves it up to the implementation. Currently, ACP has some ad-hoc cancellation mechanisms for specific features (like prompt turn cancellation via `session/cancel`), but lacks a general-purpose, per-request cancellation mechanism.
+
+This creates the following inconveniences:
+
+- cancellation should be handled for each feature separately
+- some languages that support handy cancellation mechanisms (C#, Kotlin, etc.) can't implement general-purpose request cancellation using ACP low-level machinery, and rather developers should manually call per-feature cancellation methods
+
+## What we propose to do about it
+
+Implement an **optional** `$/cancelRequest` notification method (inspired by the Language Server Protocol) that uses JSON-RPC 2.0 notification format, allowing either party (client or agent) to cancel any outstanding request by its ID.
+
+The mechanism will be:
+
+- **Optional**: Declared through capabilities exchange during initialization
+- **Unavailable** for `initialize` request (which happens before capabilities exchange)
+- **Flexible**: Provides two response options when cancellation is received:
+  1. An error response with the standard cancellation error code (-32800)
+  2. A valid response with partial or cancelled data (when meaningful partial results exist)
+
+This approach balances flexibility with standardization, allowing implementations to opt-in to cancellation support while providing predictable behavior when enabled.
+
+## Shiny future
+
+Once implemented, this enables:
+
+- **SDK integration layer**: Default mechanism that ACP SDKs can automatically wire to native language cancellation (C# CancellationToken, Kotlin Job, Go context.Context, JavaScript AbortController, etc.)
+- Individual JSON-RPC request cancellation without affecting other concurrent requests
+- Universal fallback for any request when feature-specific cancellation methods don't exist
+- Consistent cancellation behavior from both external `$/cancelRequest` and internal cancellation triggers
+- Standard error response (`-32800`) or partial results when requests are cancelled regardless of cancellation source
+
+The mechanism complements existing `session/cancel` for prompt turns while providing granular per-request control.
+
+## Implementation details and plan
+
+### Protocol Changes
+
+#### Capability Declaration
+
+Cancellation support must be declared during the capabilities exchange. Add new optional capability fields:
+
+```typescript
+interface CancellationCapabilities {
+  supportsRequestCancellation?: boolean; // Default: false
```

Comment:
I wonder if this could be an empty object for now? And it assumes if the object is there that you support general request cancellation? And then we can add more granular knobs later?

Or, we change this to just `request: boolean`